File: packages/shared-types/src/transformation.ts
export type FieldType =
  | "string"
  | "number"
  | "boolean"
  | "date"
  | "array"
  | "object";

export type Operation =
  | "uppercase"
  | "lowercase"
  | "concat"
  | "add"
  | "subtract"
  | "multiply"
  | "divide"
  | "sum_by_group"
  | "count_by_group"
  | "avg_by_group"
  | "min_by_group"
  | "max_by_group"
  | "map"
  | "filter"
  | "reduce";

export interface FieldRef {
  field?: string;
  literal?: string;
  field_type?: FieldType;
  transformations?: TransformationStep[];
}

export interface TransformationStep {
  op: Operation;
  source_field?: string;
  operand?: number;
  fields?: FieldRef[];       // for concat
  group_field?: string;      // for aggregation
  value_field?: string;      // for aggregation
  expression?: string;       // for array ops
}

export interface FieldTransformation {
  output_field: string;
  field_type: FieldType;
  pipeline: TransformationStep[];
}

export interface ValidatePipelineRequest {
  field_type: FieldType;
  pipeline: TransformationStep[];
}

export interface ValidatePipelineResponse {
  is_valid: boolean;
  errors: string[];
}

export interface PreviewPipelineRequest {
  field_type: FieldType;
  pipeline: TransformationStep[];
  sample_data: Record<string, unknown>;
}

export interface PreviewPipelineResponse {
  result: unknown;
  errors: string[];
}

File: apps/backend/src/docuweave/domain/entities/transformation.py
from dataclasses import dataclass, field
from typing import Any

from docuweave.domain.value_objects.field_type import FieldType


@dataclass(frozen=True)
class TransformationStep:
    op: str
    # For concat: list of field references (each may have nested transformations)
    fields: tuple[dict[str, Any], ...] = field(default_factory=tuple)
    # For aggregation ops
    group_field: str = ""
    value_field: str = ""
    # For math ops
    operand: float | None = None
    # For array ops
    expression: str = ""


@dataclass(frozen=True)
class FieldTransformation:
    """A single output field with its type and transformation pipeline."""
    output_field: str
    field_type: FieldType
    pipeline: tuple[TransformationStep, ...] = field(default_factory=tuple)
